client/finality-grandpa/src/finality_proof.rs

```diff
@@ -99,460 +77,226 @@ impl<BE, Block: BlockT> AuthoritySetForFinalityProver<Block> for Arc<dyn Storage
 	}
 }
 
-/// The effects of block finality.
-#[derive(Debug, PartialEq)]
-pub struct FinalityEffects<Header: HeaderT> {
-	/// The (ordered) set of headers that could be imported.
-	pub headers_to_import: Vec<Header>,
-	/// The hash of the block that could be finalized.
-	pub block: Header::Hash,
-	/// The justification for the block.
-	pub justification: Vec<u8>,
-	/// New authorities set id that should be applied starting from block.
-	pub new_set_id: u64,
-	/// New authorities set that should be applied starting from block.
-	pub new_authorities: AuthorityList,
-}
-
-/// Single fragment of proof-of-finality.
-///
 /// Finality for block B is proved by providing:
 /// 1) the justification for the descendant block F;
 /// 2) headers sub-chain (B; F] if B != F;
-/// 3) proof of GRANDPA::authorities() if the set changes at block F.
 #[derive(Debug, PartialEq, Encode, Decode, Clone)]
-pub struct FinalityProofFragment<Header: HeaderT> {
+pub struct FinalityProof<Header: HeaderT> {
 	/// The hash of block F for which justification is provided.
 	pub block: Header::Hash,
 	/// Justification of the block F.
 	pub justification: Vec<u8>,
-	/// The set of headers in the range (U; F] that we believe are unknown to the caller. Ordered.
+	/// The set of headers in the range (B; F] that we believe are unknown to the caller. Ordered.
 	pub unknown_headers: Vec<Header>,
-	/// Optional proof of execution of GRANDPA::authorities() at the `block`.
-	pub authorities_proof: Option<StorageProof>,
 }
 
-/// Proof of finality is the ordered set of finality fragments, where:
-/// - last fragment provides justification for the best possible block from the requested range;
-/// - all other fragments provide justifications for GRANDPA authorities set changes within requested range.
-type FinalityProof<Header> = Vec<FinalityProofFragment<Header>>;
-
-/// Finality proof request data.
-#[derive(Debug, Encode, Decode)]
-enum FinalityProofRequest<H: Encode + Decode> {
-	/// Original version of the request.
-	Original(OriginalFinalityProofRequest<H>),
-}
-
-/// Original version of finality proof request.
-#[derive(Debug, Encode, Decode)]
-struct OriginalFinalityProofRequest<H: Encode + Decode> {
-	/// The authorities set id we are waiting proof from.
-	///
-	/// The first justification in the proof must be signed by this authority set.
-	pub authorities_set_id: u64,
-	/// Hash of the last known finalized block.
-	pub last_finalized: H,
-}
-
-/// Prepare proof-of-finality for the best possible block in the range: (begin; end].
-///
-/// It is assumed that the caller already have a proof-of-finality for the block 'begin'.
-/// It is assumed that the caller already knows all blocks in the range (begin; end].
-///
-/// Returns None if there are no finalized blocks unknown to the caller.
-pub(crate) fn prove_finality<Block: BlockT, B: BlockchainBackend<Block>, J>(
+fn prove_finality<Block, B, J>(
 	blockchain: &B,
 	authorities_provider: &dyn AuthoritySetForFinalityProver<Block>,
```

Comment:
I think this shouldn't be needed, this is why I asked above if this trait is still needed. I think this is just an hold over from the previous implementation where we didn't know for sure when a set started or ended, therefore we figured it out by iterating over blocks and checking whether the authorities changed. Now we know for sure the last block in set X is block Y, so we can just fetch it from the database and extract the justification from it without doing any further checking. The current implementation is looking into runtime state at a given block to figure authorities, which means it won't work for old blocks if state is pruned. 

Either way I won't block this PR on this. We can deal with these changes in a follow-up PR.

Reply:
Created a new branch that I can raise as a follow-up PR after this is merged